### PR TITLE
Adapt gardener adminkubeconfig in local_setup script 

### LINKED ISSUE #844: Adapt gardener adminkubeconfig in local_setup script
**How to categorize this issue?**

/area dev-productivity
/area testing
/kind enhancement
/priority 3

**What would you like to be added**:
Static admin kube configs has been disabled from k8s version 1.27 onwards. Instead one needs to create a short-lived `adminkubeconfig` via [g/g-instructions](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md#shootsadminkubeconfig-subresource)

### PR DESCRIPTION
**Pull request by unmarshall**
**What this PR does / why we need it**:
Changes `local_setup.sh` script which has now been refactored and introduces the following enhancements:
- New Usage:
  ```bash
      Usage: local_setup.sh [Options]
      Options:
        -d | --seed                       <seed-cluster-name>                 (Required) Name of the Gardener Seed Cluster
        -t | --shoot                      <shoot-cluster-name>                (Required) Name of the Gardener Shoot Cluster
        -p | --project                    <project-name>                      (Required) Name of the Gardener Project
        -i | --provider                   <provider-name>                     (Required) Infrastructure provider name. Supported providers (gcp|aws|azure|vsphere|openstack|alicloud|metal|equinix-metal)
        -l | --landscape                  <landscape-name>                    (Optional) Name of the landscape. Defaults to dev
        -k | --kubeconfig-path           <absolute-kubeconfig-path>           (Optional) Target path to store the control and target KubeConfigs. Default: <PROJECT-DIR>/dev/kube-configs
        -m | --mcm-provider-project-path <absolute-mcm-provider-project-dir>  (Optional) MCM Provider project directory. If not provided then it assumes that both mcm and mcm-provider projects are under the same parent directory
        -e | --kubeconfig-expiry-seconds <expiry-duration-in-seconds>         (Optional) Common expiry durations in seconds for control and target KubeConfigs. Default: 3600 seconds
  ```
- Generates short-lived admin kubeconfigs.
- Introduced another script `local_restore.sh` which scales back MCM deployment to 1 and also removes the annotation that prevents DWD to scale MCM down. 
  ```bash
     Usage: local_restore.sh [Options]
    Options:
      -t | --shoot                      <shoot-cluster-name>                (Required) Name of the Gardener Shoot Cluster
      -p | --project                    <project-name>                      (Required) Name of the Gardener Project
      -l | --landscape                  <landscape-name>                    (Optional) Name of the landscape. Defaults to dev
      -i | --provider                   <provider-name>                     (Required) Infrastructure provider name. Supported providers (gcp|aws|azure|vsphere|openstack|alicloud|metal|equinix-metal)
      -k | --kubeconfig-path            <relative-kubeconfig-path>          (Optional) Relative path to the <PROJECT-DIR> where kubeconfigs will be downloaded. Path should not start with '/'. Default: <PROJECT-DIR>/dev/kube-configs
      -m | --mcm-provider-project-path  <absolute-mcm-provider-project-dir> (Optional) MCM Provider project directory. If not provided then it assumes that both mcm and mcm-provider projects are under the same parent directory
  ```
- Added a new target in Makefile to add license headers. Quite a lot of files are missing them.

**Which issue(s) this PR fixes**:
Fixes #844 

**Special notes for your reviewer**:

**Release note**:

```other developer
Developers with gardener dashboard access can now use an enhanced local_setup.sh and local_restore.sh scripts when starting MCM locally
```

```other developer
A new make target is introduced to add license headers.
```

### CODE REVIEW

File: hack/local_setup.sh [hunk 1]
```diff
@@ -1,150 +1,224 @@
...
-     exportVariables
-     updateProviderMakefile
+##############################################################################################################
+# Script sets up local development environment enabling you to start MCM process locally.
+# It does the following:
+# 1. Downloads short-lived control and target cluster kube-configs
+# 2. Ensures that kube-configs are copied to both mcm and provider-mcm project directory
+# 3. Scales down MCM to 0
+# 4. Places annotation dependency-watchdog.gardener.cloud/ignore-scaling on MCM deployment, thus preventing
+#    DWD from scaling it back up.
```
Comment (himanshu-kun) [inline hack/local_setup.sh:27]:
```suggestion
#    DWD(if deployed) from scaling it back up.
```

File: hack/local_setup.sh [hunk 2]
```diff
@@ -1,150 +1,224 @@
...
+  VIRTUAL_GARDEN_CLUSTER="sap-landscape-${LANDSCAPE_NAME}"
 }
 
-copyFilesToProvider() {
-     # Copy both the kubeconfigs to dev folder of provider 
-
-     cp "${KUBECONFIG_PATH}"/kubeconfig_target.yaml "${PROVIDER_KUBECONFIG_PATH}"/kubeconfig_target.yaml
-     cp "${KUBECONFIG_PATH}"/kubeconfig_control.yaml "${PROVIDER_KUBECONFIG_PATH}"/kubeconfig_control.yaml
+function prepare_paths() {
+  echo "Creating directory ${KUBECONFIG_PATH} if it does not exist..."
```
Comment (himanshu-kun) [inline hack/local_setup.sh:133]: thanks this was much needed :)

File: Makefile [hunk 3]
```diff
@@ -44,7 +44,7 @@ download-kubeconfigs:
 	read SHOOT; \
 	echo "enter cluster provider(gcp|aws|azure|vsphere|openstack|alicloud|metal|equinix-metal)"; \
 	read PROVIDER; \
-	./hack/local_setup.sh --SEED $$SEED --SHOOT $$SHOOT --PROJECT $$PROJECT --PROVIDER $$PROVIDER
+	./hack/local_setup.sh --seed $$SEED --shoot $$SHOOT --project $$PROJECT --provider $$PROVIDER
 
 .PHONY: local-mcm-up
 local-mcm-up: download-kubeconfigs
```
Comment (himanshu-kun) [inline Makefile:50]: I don't see any usage of `local_restore.sh`. I suppose you wanted to use it in `make local-mcm-down` ?

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
Thanks for the changes and additions, script usage looks much better now (will try it as well)

Didn't know we had so many files without licenses XD (or does it update the licenses also?)
will give lgtm after trying once (reviewer responsibility :p)

File: Makefile (hunk 3 above)
Comment (unmarshall) [reply Makefile:50]: should i create a separate make target for this?
Previously what this script does was only mentioned as comments in local_setup.sh. Now its a separate script.

Comment (himanshu-kun) [reply Makefile:50]: we have a make target by the name `local-mcm-down` for what this script is doing

Comment (unmarshall) [reply Makefile]: As discussed i have now made changes. PTAL.

**Comment by unmarshall** [comment]
As discussed @himanshu-kun will create a separate issue to update all providers with the following changes:
1. Add an empty .env file (copy it from mcm)
2. Adapt Makefile (include .env and remove the statically defined variables)

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Comment by himanshu-kun** [comment]
Updated the changes made here in https://github.com/gardener/machine-controller-manager/pull/852

**Pull request merged**